server/src/main/java/org/elasticsearch/search/fetch/subphase/FetchFieldsPhase.java

```diff
@@ -57,13 +57,28 @@ public FetchSubPhaseProcessor getProcessor(FetchContext fetchContext) {
             return null;
         }
 
+        // NOTE: FieldFetcher for non-metadata fields, as well as `_id` and `_source`.
+        // We need to retain `_id` and `_source` here to correctly populate the `StoredFieldSpecs` created by the
+        // `FieldFetcher` constructor.
         final SearchExecutionContext searchExecutionContext = fetchContext.getSearchExecutionContext();
-        final FieldFetcher fieldFetcher = fetchFieldsContext == null ? null
-            : fetchFieldsContext.fields() == null ? null
-            : fetchFieldsContext.fields().isEmpty() ? null
-            : FieldFetcher.create(searchExecutionContext, fetchFieldsContext.fields());
+        final FieldFetcher fieldFetcher = (fetchFieldsContext == null
```

Comment:
Note that here we are basically doing the opposite of what we do later with the second `FieldFetcher`. This is because the first one needs to handle all fields not handled by the second one without missing any field but also avoid including one or more fields in both.

Reply:
The goal of this logic is to keep all non metadata field that are stored and _id and _source?
Why are we skipping stored meta fields here?

Reply:
Because those are handled somewhere else using `StoredFieldSpec`. Fetching normal fields is done by means of `ValueFetcher` excluding some special ones we always want like `_id` and `_source`.

Reply:
For the other question see

```
public void process(HitContext hitContext) throws IOException {
                final Map<String, DocumentField> fields = fieldFetcher != null
                    ? fieldFetcher.fetch(hitContext.source(), hitContext.docId())
                    : Collections.emptyMap();
                final Map<String, DocumentField> metadataFields = metadataFieldFetcher.fetch(hitContext.source(), hitContext.docId());
                hitContext.hit().addDocumentFields(fields, metadataFields);
            }
```

we use two `FieldFetcher` instance...one for document fields (`fieldFetcher`) and one for metadata fields `metadataFieldFetcher`.